core/silkworm/state/in_memory_state.cpp

```diff
@@ -69,69 +67,60 @@ uint64_t InMemoryState::previous_incarnation(const evmc::address& address) const
 
 std::optional<BlockHeader> InMemoryState::read_header(uint64_t block_number,
                                                       const evmc::bytes32& block_hash) const noexcept {
-    if (block_number >= headers_.size()) {
-        return std::nullopt;
-    }
-
-    auto it{headers_[block_number].find(block_hash)};
-    if (it == headers_[block_number].end()) {
-        return std::nullopt;
+    const auto it1{headers_.find(block_number)};
+    if (it1 != headers_.end()) {
+        const auto it2{it1->second.find(block_hash)};
+        if (it2 != it1->second.end()) {
+            return it2->second;
+        }
     }
-    return it->second;
+    return std::nullopt;
 }
 
 std::optional<BlockBody> InMemoryState::read_body(uint64_t block_number,
                                                   const evmc::bytes32& block_hash) const noexcept {
-    if (block_number >= bodies_.size()) {
-        return std::nullopt;
-    }
-
-    auto it{bodies_[block_number].find(block_hash)};
-    if (it == bodies_[block_number].end()) {
-        return std::nullopt;
+    const auto it1{bodies_.find(block_number)};
+    if (it1 != bodies_.end()) {
+        const auto it2{it1->second.find(block_hash)};
+        if (it2 != it1->second.end()) {
+            return it2->second;
+        }
     }
-    return it->second;
+    return std::nullopt;
```

Comment:
I would prefer `return {}` rather than `return std::nullopt`.

Reply:
IMHO `std::nullopt` is clearer.